Fixed issue: Comment notification not working after implementing OAuth2 and Added PKCE code for OAuth2

raghavaggarwal2308 opened the pull request:
### Summary
1. Comment webhooks not working: Comment webhook notifications were not working because earlier Jira wasn't sending info about the issue on which the comment was posted. So, we were using a dummy JWT to fetch that info separately. Since we are using OAuth2 now, that part of the code was giving an error. We explored a bit and found out that issue info is now sent in the [webhook payload for comments](https://developer.atlassian.com/cloud/jira/platform/change-notice-removal-of-comments-from-issue-webhooks). So, we removed it.
2. Added PKCE along with authorization code while performing OAuth2.
3. Did some code refactoring.

### Issue/Enhancement Links:
1. [Add PKCE while performing OAuth](https://github.com/mattermost/mattermost-plugin-jira/pull/949#discussion_r1221636177)
2. [Webhook notification not working](https://github.com/mattermost/mattermost-plugin-jira/pull/949#issuecomment-1579493594)

### Fixes PR
#949 (Implement OAuth2 Authentication (reopened))

mattermost-build commented:
Hello @raghavaggarwal2308,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

DHaussermann commented:
@raghavaggarwal2308 FWIW - I'll try and do a test from this PR branch and see if the subscription bug is solved. 

Maybe @hanzei can help you push this back up into https://github.com/mattermost/mattermost-plugin-jira/pull/949 shortly.

DHaussermann commented:
@raghavaggarwal2308 I'm a bit confused. I compiled this branch and upgraded the same server where I found the issue and I see no change in behavior.  Updating several fields in the ticket is delivering the events but comment events do not get delivered. 

I double checked my commit
![image](https://github.com/mattermost/mattermost-plugin-jira/assets/35588783/a3a2fde3-9970-4ff9-a052-0942221f42ef)
 
 ```
 {
  "caller": "app/plugin_api.go:976",
  "level": "error",
  "msg": "PostNotifications: error while getting jiraClient, err: failed to get Jira client for the user : error in getting token from token source: oauth2: token expired and refresh token is not set",
  "plugin_id": "jira",
  "timestamp": "2023-06-15 16:48:34.765 Z"
}
```
 The error is now different. I don't know what token we're looking for here.
 
 I also tried on a fresh server with a new installation and comments still don't work.
 
 Please let me know if you can repro this.

esarafianou reviewed (approved):
LGTM

raghavaggarwal2308 commented:
@DHaussermann We tried to reproduce this on MM server version v7.8.2 but we were unable to do so. Can you please share the server version you are using? Also, it would be helpful if you can share your Mattermost and Jira test cloud server credentials so that we can adequately debug the issue here.

DHaussermann commented:
Hi @raghavaggarwal2308, 

I will DM you a set of credentials now. I can also include an invite to a server where I see the issue. 
Please ping me if you still cannot repo after.

DHaussermann commented:
@raghavaggarwal2308 after re-testing this. I do see comment events being delivered after the last commit. 

The issue I see now is that unlike other events, comment events don't work with security levels for me. (On a project without security levels, this would not be something you can find as the case is not applicable)

Scenario A - The Jira issue has no security level set
- Create a subscription that targets a project with security levels such as `DKH-Project`
- Add events including comment events
- Do **not** add security level filter to your subscription
- Create or locate an issue in the project
- Remove the security level (If it was applied at time of creation)
- Add or update a comment
**Observed** Events are delivered

Scenario B - The Jira issue has a security level set
- Create a subscription that targets a project with security levels such as `DKH-Project`
- Add events including comment events
- Add security level filter to your subscription (can be all of them so any security level will deliver)
- Create or locate an issue in the project
- Add a security level (If it was not already applied at time of creation)
- Add or update a comment
**Observed** Events are delivered but other filed changes get event deliveries as expected

raghavaggarwal2308 commented:
@DHaussermann @mickmister We explored the above issue. The reason for that is the webhook body for comment notification contains the issue details now, but they are incomplete. The custom/additional fields present on the issue are not listed in the issue details. So, when we are unable to match the filters for comment notifications when the "Security level" filter is set. To resolve this issue we need to make an extra API call to fetch the issue details separately for comment notification, but we need either the user's JiraAccountID or the MattermostUserID to get the Jira client and make an API call. The problem is that the webhook body contains only the account ID of the commenter, but that user may or may not be connected to Mattermost, so if we use that ID we won't be able to get issue details in case the user is not connected and the notification will not be posted. Can you suggest us a way to get the connected user's ID?

DHaussermann commented:
@jasonblais heads-up Michael is likely the person with the most knowledge to provide guidance here but he's still away.

@hanzei may have some context but has not work on this functionality as much.

hanzei commented:
I don't have context knowledge about the webhook handling in JIRA to share.

mickmister commented:
> The reason for that is the webhook body for comment notification contains the issue details now, but they are incomplete. The custom/additional fields present on the issue are not listed in the issue details. So, when we are unable to match the filters for comment notifications when the "Security level" filter is set. 

> The problem is that the webhook body contains only the account ID of the commenter, but that user may or may not be connected to Mattermost, so if we use that ID we won't be able to get issue details in case the user is not connected and the notification will not be posted. Can you suggest us a way to get the connected user's ID?

@raghavaggarwal2308 This may be a limitation we need to accept. Can you please provide what information is missing from the payload, and what is present? Is there any way to have a superuser token / service account for Jira Cloud? If not, we may need to be clear that "any subscription post made from a comment will need to be made by a connected user."

raghavaggarwal2308 commented:
@mickmister
1. According to what I have explored, I don't think Jira supports service account creation.
2. In the webhook body for comment notification only the default fields are present like summary, issue type, project, assignee, priority, and status. The custom fields like security level, time tracking, etc. are not sent in the body. 

Comment webhook body:
```
{
  "issue": {
    "id": "10080",
    "self": "https://brightscout-dev.atlassian.net/rest/api/2/10080",
    "key": "TEST-68",
    "fields": {
      "summary": "Test-issue-summary",
      "issuetype": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/issuetype/10001",
        "id": "10001",
        "description": "Created by Jira Agile - do not edit or delete. Issue type for a user story.",
        "iconUrl": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10315?size=medium",
        "name": "Story",
        "subtask": false,
        "avatarId": 10315,
        "hierarchyLevel": 0
      },
      "project": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/project/10000",
        "id": "10000",
        "key": "TEST",
        "name": "Test",
        "projectTypeKey": "software",
        "simplified": false,
        "avatarUrls": {
          "48x48": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424",
          "24x24": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424?size=small",
          "16x16": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424?size=xsmall",
          "32x32": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424?size=medium"
        }
      },
      "assignee": null,
      "priority": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/priority/3",
        "iconUrl": "https://brightscout-dev.atlassian.net/images/icons/priorities/medium.svg",
        "name": "Medium",
        "id": "3"
      },
      "status": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/status/10000",
        "description": "",
        "iconUrl": "https://brightscout-dev.atlassian.net/",
        "name": "Backlog",
        "id": "10000",
        "statusCategory": {
          "self": "https://brightscout-dev.atlassian.net/rest/api/2/statuscategory/2",
          "id": 2,
          "key": "new",
          "colorName": "blue-gray",
          "name": "To Do"
        }
      }
    }
  }
}
```

Issue details in issue created webhook body:
```
{
  "issue": {
    "id": "10080",
    "self": "https://brightscout-dev.atlassian.net/rest/api/2/10080",
    "key": "TEST-68",
    "fields": {
      "statuscategorychangedate": "2023-07-05T04:39:41.891-0500",
      "issuetype": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/issuetype/10001",
        "id": "10001",
        "description": "Functionality or a feature expressed as a user goal.",
        "iconUrl": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10315?size=medium",
        "name": "Story",
        "subtask": false,
        "avatarId": 10315,
        "hierarchyLevel": 0
      },
      "timespent": null,
      "customfield_10030": null,
      "project": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/project/10000",
        "id": "10000",
        "key": "TEST",
        "name": "Test",
        "projectTypeKey": "software",
        "simplified": false,
        "avatarUrls": {
          "48x48": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424",
          "24x24": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424?size=small",
          "16x16": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424?size=xsmall",
          "32x32": "https://brightscout-dev.atlassian.net/rest/api/2/universal_avatar/view/type/project/avatar/10424?size=medium"
        }
      },
      "customfield_10033": [],
      "fixVersions": [],
      "aggregatetimespent": null,
      "customfield_10034": null,
      "customfield_10035": null,
      "resolution": null,
      "customfield_10036": null,
      "customfield_10037": null,
      "customfield_10027": null,
      "customfield_10028": null,
      "customfield_10029": null,
      "resolutiondate": null,
      "workratio": -1,
      "issuerestriction": {
        "issuerestrictions": {},
        "shouldDisplay": false
      },
      "lastViewed": null,
      "watches": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/issue/TEST-68/watchers",
        "watchCount": 0,
        "isWatching": false
      },
      "created": "2023-07-05T04:39:41.497-0500",
      "customfield_10020": null,
      "customfield_10021": null,
      "customfield_10022": null,
      "customfield_10023": null,
      "priority": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/priority/3",
        "iconUrl": "https://brightscout-dev.atlassian.net/images/icons/priorities/medium.svg",
        "name": "Medium",
        "id": "3"
      },
      "customfield_10024": null,
      "customfield_10025": null,
      "customfield_10026": null,
      "labels": [],
      "customfield_10016": null,
      "customfield_10017": null,
      "customfield_10018": {
        "hasEpicLinkFieldDependency": false,
        "showField": false,
        "nonEditableReason": {
          "reason": "PLUGIN_LICENSE_ERROR",
          "message": "The Parent Link is only available to Jira Premium users."
        }
      },
      "customfield_10019": "0|i000h3:",
      "aggregatetimeoriginalestimate": null,
      "timeestimate": null,
      "versions": [],
      "issuelinks": [],
      "assignee": null,
      "updated": "2023-07-05T04:39:41.497-0500",
      "status": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/status/10000",
        "description": "",
        "iconUrl": "https://brightscout-dev.atlassian.net/",
        "name": "Backlog",
        "id": "10000",
        "statusCategory": {
          "self": "https://brightscout-dev.atlassian.net/rest/api/2/statuscategory/2",
          "id": 2,
          "key": "new",
          "colorName": "blue-gray",
          "name": "New"
        }
      },
      "components": [],
      "customfield_10050": null,
      "customfield_10051": null,
      "timeoriginalestimate": null,
      "customfield_10052": null,
      "description": null,
      "customfield_10010": null,
      "customfield_10014": null,
      "customfield_10015": null,
      "timetracking": {},
      "customfield_10005": null,
      "customfield_10006": null,
      "security": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/securitylevel/10001",
        "id": "10001",
        "description": "",
        "name": "TEST_SECURITY_LEVEL"
      },
      "customfield_10007": null,
      "customfield_10008": null,
      "aggregatetimeestimate": null,
      "customfield_10009": null,
      "attachment": [],
      "summary": "Test-issue-summary",
      "creator": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/user?accountId=6282e44146f77e006f20c3d9",
        "accountId": "6282e44146f77e006f20c3d9",
        "avatarUrls": {
          "48x48": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png",
          "24x24": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png",
          "16x16": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png",
          "32x32": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png"
        },
        "displayName": "Raghav Aggarwal",
        "active": true,
        "timeZone": "America/Chicago",
        "accountType": "atlassian"
      },
      "subtasks": [],
      "customfield_10040": null,
      "customfield_10041": null,
      "customfield_10042": null,
      "reporter": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/user?accountId=6282e44146f77e006f20c3d9",
        "accountId": "6282e44146f77e006f20c3d9",
        "avatarUrls": {
          "48x48": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png",
          "24x24": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png",
          "16x16": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png",
          "32x32": "https://secure.gravatar.com/avatar/010054b5381598d1f6e68b4eed5a0550?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FRA-2.png"
        },
        "displayName": "Raghav Aggarwal",
        "active": true,
        "timeZone": "America/Chicago",
        "accountType": "atlassian"
      },
      "customfield_10044": null,
      "aggregateprogress": {
        "progress": 0,
        "total": 0
      },
      "customfield_10045": null,
      "customfield_10001": null,
      "customfield_10046": null,
      "customfield_10002": null,
      "customfield_10047": null,
      "customfield_10003": null,
      "customfield_10004": null,
      "customfield_10048": null,
      "customfield_10038": null,
      "customfield_10039": null,
      "environment": null,
      "duedate": null,
      "progress": {
        "progress": 0,
        "total": 0
      },
      "votes": {
        "self": "https://brightscout-dev.atlassian.net/rest/api/2/issue/TEST-68/votes",
        "votes": 0,
        "hasVoted": false
      }
    }
  }
}
```

mickmister commented:
@raghavaggarwal2308 In that case, I think we'll need to accept the fact that any subscription post made from a comment will need to be made by a connected user. DM mentions should still work in this case though right?

raghavaggarwal2308 commented:
@mickmister We tested the DM notifications and turns out they are not working after the oauth changes. After debugging we found that we are making an API call to fetch the comment details and validate whether the user has permission to see the comment or not. We are making the API call on the comment URL present in the webhook payload (Ex: https://your-domain.atlassian.net/rest/api/2/issue/{issue-id}/comment/{comment-id}). But according to the [Jira Documentation](https://developer.atlassian.com/cloud/jira/platform/oauth-2-3lo-apps/#3-2-construct-the-request-url), the request should be made via `api.atlassian.com` and not `https://your-domain.atlassian.net` due to which the API call is getting failed. We think the other API might also fail because of this.

mickmister commented:
@raghavaggarwal2308 Can we make a post like "A new comment has been posted on (issue)" as a fallback? Also, how is it going on this issue in general?

raghavaggarwal2308 commented:
> @raghavaggarwal2308 Can we make a post like "A new comment has been posted on (issue)" as a fallback? Also, how is it going on this issue in general?

@mickmister I am a bit confused about where we need to make the change. We have two issues here:

1. Comment notification not working when a non-connected user adds a comment on Jira issue [(Due to missing details in comment webhook payload)](https://github.com/mattermost/mattermost-plugin-jira/pull/953#issuecomment-1604073508): In this case if we add a fallback, we will end up notification even in the cases when subscription filter are not matched.

2. DM mentions not working [(Because of invalid API call)](https://github.com/mattermost/mattermost-plugin-jira/pull/953#issuecomment-1624116221): In this case if we add a fallback, we will be sending notification without validating that the user has permission to access the issue or not. I think that might be wrong.

In general, whenever we get a webhook request from Jira, we match the subscription filters and post a notification in the channel:
![comment-notification](https://github.com/mattermost/mattermost-plugin-jira/assets/55234496/8f41334d-3b7e-4d09-9b1e-8459498b75db)

mickmister commented:
> 1. Comment notification not working when a non-connected user adds a comment on Jira issue [(Due to missing details in comment webhook payload)](https://github.com/mattermost/mattermost-plugin-jira/pull/953#issuecomment-1604073508): In this case if we add a fallback, we will end up notification even in the cases when subscription filter are not matched.

> 2. DM mentions not working [(Because of invalid API call)](https://github.com/mattermost/mattermost-plugin-jira/pull/953#issuecomment-1624116221): In this case if we add a fallback, we will be sending notification without validating that the user has permission to access the issue or not. I think that might be wrong.

@raghavaggarwal2308 Right, I was mistaken. The issue is with the subscription filtering so there's nothing that can be done there.

For the invalid API call related to the DM notification feature, is this something we can correct by using `api.atlassian.com`?

mickmister commented:
> Do we need to make this change for all the APIs or just for the DM comment notifications?

@raghavaggarwal2308 Sorry I don't know the answer to this. I would say if Atlassian recommends this, and it all works when making the change, then we should just change all of them.

raghavaggarwal2308 commented:
@mickmister Fixed both issues with the comment webhook notification:

1. [Comment webhook notification not working for comment webhook notifications when "Security Level" filter is applied in subscription](https://github.com/mattermost/mattermost-plugin-jira/pull/953#issuecomment-1604073508): Updated the logic to expand the issue details in the webhook notification. NOTE: The details will be expanded only when the [comment is made by a connected user](https://github.com/mattermost/mattermost-plugin-jira/pull/953#issuecomment-1621988500).

2. [DM notification not working for comment webhook notifications](https://github.com/mattermost/mattermost-plugin-jira/pull/953#issuecomment-1624116221): Updated the code to make API call with proper base URL.

mickmister commented on server/webhook_jira.go:
```diff
@@ -29,6 +31,50 @@ type JiraWebhook struct {
+		return err
+	}
+
+	// Jira Cloud comment event. We need to fetch issue data because it is not expanded in webhook payload.
+	isCommentEvent := jwh.WebhookEvent == commentCreated || jwh.WebhookEvent == commentUpdated || jwh.WebhookEvent == commentDeleted
+	if isCommentEvent && instance.Common().IsCloudInstance() {
```
I think we should check if it's a cloud instance first and return if Jira server, to show that this function doesn't care about Jira server first and foremost

mickmister commented on server/webhook_jira.go:
```diff
@@ -29,6 +31,50 @@ type JiraWebhook struct {
+		} else if _, ok := instance.(*cloudOAuthInstance); ok {
+			mmUserID, err := p.userStore.LoadMattermostUserID(instanceID, jwh.Comment.Author.AccountID)
+			if err != nil {
+				return err
```
I think we should return a more specific error here. Something like:

```suggestion
				return errors.Wrap(err, "Jira comment author is not connected to Mattermost. Cannot create subscription posts for this comment.")
```

mickmister commented on server/webhook.go:
```diff
@@ -140,7 +140,7 @@ func (wh *webhook) PostNotifications(p *Plugin, instanceID types.ID) ([]*model.P
 		isCommentEvent := wh.Events().Intersection(commentEvents).Len() > 0
 		if isCommentEvent {
-			err = client.RESTGet(notification.commentSelf, nil, &struct{}{})
+			err = client.RESTGet(fmt.Sprintf("/2/issue/%s/comment/%s", wh.Issue.ID, wh.Comment.ID), nil, &struct{}{})
```
Wondering if this works for all 3 Jira instance types

raghavaggarwal2308 replied:
@mickmister I have tested this code on Jira cloud-oauth and server instances but since the Jira JWT is not working for us, we were not able to test it there.

raghavaggarwal2308 replied:
@mickmister Added the conditional block

mickmister commented on server/issue.go:
```diff
@@ -736,6 +736,30 @@ func getIssueCustomFieldValue(issue *jira.Issue, key string) StringSet {
 	return nil
 }
 
+func (p *Plugin) getIssueDataForCloudWebhook(instance Instance, issueKey string) (*jira.Issue, error) {
+	ci, ok := instance.(*cloudInstance)
+	if !ok {
+		return nil, errors.Errorf("must be a Jira cloud instance, is %s", instance.Common().Type)
+	}
+
+	jiraClient, err := ci.getClientForBot()
+	if err != nil {
+		return nil, err
+	}
+
+	issue, resp, err := jiraClient.Issue.Get(issueKey, nil)
+	if err != nil {
+		switch {
+		case resp == nil:
+			return nil, errors.WithMessage(userFriendlyJiraError(nil, err), "request to Jira failed")
+		case resp.StatusCode == http.StatusNotFound || resp.StatusCode == http.StatusUnauthorized:
+			return nil, errors.New(`we couldn't find the issue key, or the cloud "bot" client does not have the appropriate permissions to view the issue`)
+		}
+	}
+
+	return issue, nil
+}
+
```
Did this function exist somewhere else earlier? Was it deleted at some point? I'm asking because there is no red diff related to this function in this commit

raghavaggarwal2308 replied:
@mickmister Yes, it was removed earliar

mickmister reviewed (approved):
LGTM, though I want to make sure the comment fetching works for Jira Server on this line https://github.com/mattermost/mattermost-plugin-jira/blob/0186841afa2f8b1de2efcfe3177145dc661ebeb5/server/webhook.go#L143

raghavaggarwal2308 commented:
@mickmister FIxed the review comments mentioned by you

raghavaggarwal2308 commented:
@DHaussermann Are you testing this PR? If not, can BrightScout test it?

Kshitij-Katiyar commented:
> @raghavaggarwal2308: @DHaussermann is currently on vacation so he won't be able to respond
> 
> Also there's an unresolved discussion here [#953 (comment)](https://github.com/mattermost/mattermost-plugin-jira/pull/953#discussion_r1273592880). I'm thinking we should keep it the way it was for Jira Server, to avoid having to test with Jira Server v7. However, Jira Server v9 should be regression tested before OAuth is released.

@mickmister Sure, we will be making this today after that as @DHaussermann is not available, so should brightscout QA do the QA review if this is urgent or should we wait for Dylan to come back ?

mickmister commented:
@Kshitij-Katiyar Brightscout can do QA on this PR. FYI Dylan will be back on Monday

mickmister commented:
@DHaussermann Note that I've merged this into the main OAuth PR https://github.com/mattermost/mattermost-plugin-jira/pull/949

AayushChaudhary0001 commented:
Tested and approved, LGTM

- This PR was tested on MM version-7.8.2 with latest Jira cloud instance and server version - 9.
- Comment notifications were found working during the testing of this PR.
- This PR is tested and the setup of OAuth 2.0 is completely working fine.
